### PR TITLE
Refactor network io

### PR DESCRIPTION
**Pull request by tallytalwar**
A Utility Class to do Unity IO

- Provides a utlity class to do network or file io (for bundled resource assets) using unity interfaces
- Uses the UnityIO interface to fetch tiles

### CODE REVIEW

File: Assets/MapzenMap.cs [hunk 1]
```diff
@@ -21,81 +20,72 @@ public class MapzenMap : MonoBehaviour
...
-            {
-                int tileX = TileX + x;
-                int tileY = TileY + y;
-
-                var url = string.Format("https://tile.mapzen.com/mapzen/vector/v1/all/{0}/{1}/{2}.json?api_key={3}",
-                              TileZ, tileX, tileY, ApiKey);
+    private UnityIO tileIO = new UnityIO();
 
-                Debug.Log("URL request " + url);
+    void fetchNetworkTile(object userData, Response response) {
```
Comment (karimnaaji) [inline Assets/MapzenMap.cs]: What do you think of calling this `onNetworkTileFetched`? At this point we don't fetch the tile, but we just received the data for it.

File: Assets/MapzenMap.cs [hunk 2]
```diff
@@ -21,81 +20,72 @@ public class MapzenMap : MonoBehaviour
...
-        yield return request.Send();
+        // Construct the HTTP request
+        for (int x = 0; x < TileRangeX; ++x)
+        {
+            for (int y = 0; y < TileRangeY; ++y)
+            {
+                int tileX = TileX + x;
+                int tileY = TileY + y;
+
+                // var url = string.Format("/Users/varuntalwar/Desktop/test/{0}/{1}/{2}.json",
```
Comment (karimnaaji) [inline Assets/MapzenMap.cs]: Leftover

File: Assets/UnityIO.cs [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+	}
+	public Response(string e, string d) {
+		error = e;
+		data = Convert.FromBase64String(d);
+	}
+}
+
+public class UnityIO 
+{
+	public IEnumerator UnityIORequest(string url, UnityIORequestCallback callback, object userData) {
```
Comment (karimnaaji) [inline Assets/UnityIO.cs]: `UnityIORequest` is a little redundant with the current class name, what about `Fetch`?

File: Assets/UnityIO.cs [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+			networkRequest = false;
+		}
+
+		if (networkRequest) {
+			WWW request = new WWW(url);
+			yield return request;
+			response = new Response(request.error, request.bytes);
+		} else {
+			yield return 0;
+			if (!File.Exists(url)) {
```
Comment (karimnaaji) [inline Assets/UnityIO.cs]: I wonder how the path resolution would on a mobile device, it might be safe to directly use the convenience Unity method https://docs.unity3d.com/ScriptReference/Resources.Load.html.

File: Assets/UnityIO.cs [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+	}
+}
+
+public class UnityIO 
+{
+	public IEnumerator UnityIORequest(string url, UnityIORequestCallback callback, object userData) {
+
+		bool networkRequest = false;
+		Response response;
+		if (url.StartsWith("http")) {
```
Comment (karimnaaji) [inline Assets/UnityIO.cs]: Do you think we can leverage the usage of the `URL` class that .NET provides? https://msdn.microsoft.com/en-us/library/system.uri(v=vs.110).aspx.

We would have those built in.

File: Assets/UnityIO.cs [hunk 6]
```diff
@@ -0,0 +1,49 @@
+using UnityEngine;
+using System;
+using System.IO;
+using System.Collections;
+
+public delegate void UnityIORequestCallback(object userData, Response response);
```
Comment (karimnaaji) [inline Assets/UnityIO.cs]: We might be able to reduce the visibility by leaving that within the `UnityIO` class scope.

File: Assets/UnityIO.cs [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+
+public class Response {
```
Comment (karimnaaji) [inline Assets/UnityIO.cs]: Same here, not sure if a nested class would make more sense, but it is possible to leave `Response` within `UnityIO` scope.

**Comment by tallytalwar** [comment]
@karimnaaji all these comments were great. Addressed all.

File: Assets/UnityIO.cs [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+			WWW request = new WWW(uri.AbsoluteUri);
+			yield return request;
+			response = new Response(request.error, request.bytes);
+		} else {
+			yield return 0;
+			var obj = Resources.Load(uri.AbsolutePath);
+			if (obj == null) {
+				response = new Response("Resource not found", "");
+			} else {
+				response = new Response(null, obj.ToString());
```
Comment (karimnaaji) [inline Assets/UnityIO.cs]: This doesn't work because `obj.ToString()` returns the name of the file.

`Resources.Load` only load asset that are Unity assets, basically in a case similar to this:

```c#
// If tileAsset is a .prefab
GameObject go = Instantiate (Resources.Load("tileAsset")) as GameObject;
```
That we can use for loading materials/textures etc.. 

For text files (for example a `.json` resource) you can simply load it like that:

```c#
TextAsset tileTextAsset = Resources.Load("mapzen-tile.json") as TextAsset;
```

We could add an interface to load Unity assets, using the proper way to load them (similar to web requests):

```c#
// 1. Construct the URI considering using Application.dataPath + "/Resources/" as base path
// 2. Load the resource async
ResourceRequest request = Resources.LoadAsync(uri.aboslutePath);
yield return request;
// 3. Construct the response
```

If we want to stick with using `byte[]` we can load file as a binary in a synchronous way like this https://github.com/tangrams/tangram-unity/commit/7c882b383f7a0ee86fe8e28d82d8517cd104abf7, feel free to cherry-pick the commit. I think it will be enough for now, we can add the interface for unity assets when we start dealing with more resources.

File: Assets/UnityIO.cs [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+		}
+		public Response(string e, string d) {
+			error = e;
+			data = Convert.FromBase64String(d);
+		}
+	}
+
+	public delegate void IORequestCallback(object userData, Response response);
+
+	public IEnumerator FetchData(Uri uri, IORequestCallback callback, object userData) {
```
Comment (matteblair) [inline Assets/UnityIO.cs]: Instead of `delegate IORequestCallback` we can declare an `interface IFileResponseHandler` - any object that needs to receive an IO file response can just implement the interface and then there's no need for a `userData` object.

File: Assets/UnityIO.cs [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+		public string error;
+		public byte[] data;
+
+		public Response(string e, byte[] d) {
+			error = e;
+			data = d;
+		}
+		public Response(string e, string d) {
+			error = e;
+			data = Convert.FromBase64String(d);
```
Comment (matteblair) [inline Assets/UnityIO.cs]: How do you imagine this being used? I would lean towards leaving data transformations like this to receiver of the response.

File: Assets/UnityIO.cs [hunk 11]
```diff
@@ -0,0 +1,48 @@
+using UnityEngine;
+using System;
+using System.IO;
+using System.Collections;
+
+public class UnityIO
+{
+	public class Response {
+		public string error;
```
Comment (matteblair) [inline Assets/UnityIO.cs]: Can you explain with comments how `error` should be interpreted? If a request is successful, is `error` set to `null`, `""`, `"Success"`, etc.

File: Assets/UnityIO.cs [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+	public IEnumerator FetchData(Uri uri, IORequestCallback callback, object userData) {
+
+		bool networkRequest = false;
+		Response response;
+		if (uri.Scheme == Uri.UriSchemeHttp || uri.Scheme == Uri.UriSchemeHttps) {
+			networkRequest = true;
+		}
+
+		if (networkRequest) {
+			WWW request = new WWW(uri.AbsoluteUri);
```
Comment (matteblair) [inline Assets/UnityIO.cs]: Any reason for using `WWW` over `UnityWebRequest`? I don't actually know how they differ, I'm curious.

Comment (tallytalwar) [inline Assets/UnityIO.cs]: In practice and in documentation I did not find any difference. Documentation mentions the use of both. `WWW` had just a bit less boiler plate code compared to `UnityWebRequest` hence I chose to use `WWW` over `UnityWebRequest`.

File: Assets/UnityIO.cs (hunk 10 above)
Comment (tallytalwar) [inline Assets/UnityIO.cs]: I wanted to keep the response as an array of raw bytes and hence this transformation.

This makes sure that the receiver of the response can transform these "bytes" to a suitable format.

File: Assets/UnityIO.cs (hunk 9 above)
Comment (tallytalwar) [inline Assets/UnityIO.cs]: Thats a good point, will check it out and update if it make sense.

File: Assets/UnityIO.cs (hunk 11 above)
Comment (tallytalwar) [inline Assets/UnityIO.cs]: Sure will do. An `empty` string means no error. But will put apt comments.

**Comment by tallytalwar** [comment]
@matteblair I agree on the generic fileIO bit. Here the work in more on network and asset resource IO instead of file IO.

File: Assets/UnityIO.cs (hunk 11 above)
Comment (matteblair) [inline Assets/UnityIO.cs]: According to https://docs.unity3d.com/ScriptReference/WWW-error.html it could be either empty or null depending on the platform. One of your code branches also sets it to null. I recommend normalizing the "success" case to always return one value, whichever makes more sense to you. And noting that value with comments, of course.

File: Assets/UnityIO.cs (hunk 10 above)
Comment (matteblair) [inline Assets/UnityIO.cs]: Let me rephrase: Why does this class need to be aware of how the response data is encoded? Decoding the response is the responsibility of the code that made the request, right?

File: Assets/UnityIO.cs (hunk 9 above)
Comment (matteblair) [inline Assets/UnityIO.cs]: We could also just do away with the `userData` argument and use lambdas or instance methods for the callback delegate. I would still prefer an interface. A `userData` object is definitely not idiomatic C# though, we should address that.

File: Assets/UnityIO.cs (hunk 12 above)
Comment (matteblair) [inline Assets/UnityIO.cs]: Judging by the docs, `UnityWebRequest` has more granular access to parameters that we will probably be interested in, like request timeout and response codes. It seems to me that we will probably want to use `UnityWebRequest` over `WWW` in the long term.

**Comment by tallytalwar** [comment]
K picking this up again now!

File: Assets/UnityIO.cs (hunk 8 above)
Comment (tallytalwar) [inline Assets/UnityIO.cs]: Thanks @karimnaaji I do not think we should have generic FileIO but it does make sense to have TextAsset IO from a bunded resource. So I added the code for that using `Resources.LoadAsync`.

File: Assets/UnityIO.cs (hunk 11 above)
Comment (tallytalwar) [inline Assets/UnityIO.cs]: Ok Addressed.

File: Assets/UnityIO.cs (hunk 10 above)
Comment (tallytalwar) [inline Assets/UnityIO.cs]: Addressed the usage of this so the transformation lands on user's hands.

File: Assets/UnityIO.cs (hunk 9 above)
Comment (tallytalwar) [inline Assets/UnityIO.cs]: Partially addressed this using lambdas. Need to see how an interface could be a better approach to handle this. But lambda usage makes this look much better.

File: Assets/UnityIO.cs (hunk 12 above)
Comment (tallytalwar) [inline Assets/UnityIO.cs]: Reverted to use `UnityWebRequest` now.

**Comment by tallytalwar** [comment]
I am mostly done here, will give some thought/reading on if providing an UnityIOHandler interface is a better approach to what I have with delegates and lambdas right now. Other than that, this is again open for reviewing.

File: Assets/MapzenMap.cs [hunk 13]
```diff
@@ -22,73 +28,60 @@ public class MapzenMap : MonoBehaviour
...
+
     void Start()
     {
-        TileBounds bounds = new TileBounds(Area);
-
-        foreach (var tileAddress in bounds.TileAddressRange)
+        // Construct the HTTP request
+        for (int x = 0; x < TileRangeX; ++x)
         {
-            var wrappedTileAddress = tileAddress.Wrapped();
```
Comment (karimnaaji) [inline Assets/MapzenMap.cs]: This has been removed probably from the rebase, we don't want to use the tile range anymore so this can go.

File: Assets/UnityIO.cs [hunk 14]
```diff
@@ -0,0 +1,76 @@
...
+                requestError = "HTTP Error with errorcode: " + request.responseCode;
+                // Following not available in 5.6.1 version
+                //requestError = UnityWebRequest.GetErrorDescription(request.responseCode);
+            }
+            response = new Response(requestError, request.downloadHandler.data);
+        }
+        callback(response);
+    }
+
+    // Note: uri must be constructed using Application.dataPath + "/Resources/" + assetName as base path
```
Comment (karimnaaji) [inline Assets/UnityIO.cs]: Should we check that explicitly from the `uri`?

File: Assets/UnityIO.cs [hunk 15]
```diff
@@ -0,0 +1,76 @@
...
+        if (uri.Scheme != Uri.UriSchemeFile) {
+            response = new Response("Wrong Uri scheme provided", null);
+        }
+        else {
+            ResourceRequest request = Resources.LoadAsync(uri.AbsolutePath);
+            yield return request;
+
+            TextAsset textAsset = request.asset as TextAsset;
+            if (textAsset == null) {
+                response = new Response("Missing asset at path: " + uri.AbsolutePath, null);
```
Comment (karimnaaji) [inline Assets/UnityIO.cs]: Do you know if `textAsset` is also `null` if the `request.asset` is anything else than a text resource?

File: Assets/MapzenMap.cs (hunk 13 above)
Comment (tallytalwar) [inline Assets/MapzenMap.cs]: Will update. Thanks 👍

**Comment by tallytalwar** [comment]
@karimnaaji addressed your comments and suggestions.
